source/common/router/rds_impl.cc

```diff
@@ -211,6 +222,37 @@ void RdsRouteConfigProviderImpl::onConfigUpdate() {
     prev_config->config_ = new_config;
     return previous;
   });
+
+  const auto aliases = config_update_info_->aliasesInLastVhdsUpdate();
+  // Regular (non-VHDS) RDS updates don't populate aliases fields in resources.
+  if (aliases.empty()) {
+    return;
+  }
+
+  // Notifies connections that RouteConfiguration update has been propagated.
+  // Callbacks processing is performed in FIFO order and stops when the alias(es) used in the VHDS
+  // update request do not match the aliases in the update response. The assumption is the response
+  // to a request is not going to arrive before all requests before it have been responded to.
+  config_update_callbacks_->runOnAllThreads(
+      [aliases](ThreadLocal::ThreadLocalObjectSharedPtr previous)
+          -> ThreadLocal::ThreadLocalObjectSharedPtr {
+        auto callbacks = std::dynamic_pointer_cast<ThreadLocalCallbacks>(previous)->callbacks_;
+        std::vector<std::string> aliases_not_in_update;
+        while (!callbacks.empty()) {
+          auto update_on_demand_callback = callbacks.front();
+          std::set_difference(update_on_demand_callback.aliases_.begin(),
+                              update_on_demand_callback.aliases_.end(), aliases.begin(),
+                              aliases.end(), std::back_inserter(aliases_not_in_update));
+          if (aliases_not_in_update.empty()) {
+            callbacks.pop();
+            update_on_demand_callback.cb_();
```

Comment:
I have some life-circle concern here, a stream may not live as long as a subscription, in which case this callback ends in unknown behavior.

Reply:
Do you mean the gRPC stream carrying the subscription? If so, it is a safe assumption; the subscription is a shared owner of a GrpcMux, which is the owner of the GrpcStream.

Reply:
no, I mean the per HTTP request ActiveStream.

Reply:
I pause the filter chain until a response arrives and the callback resumes its execution. Wouldn't that mean that the ActiveConnection sticks around too?

Reply:
No, I am saying a subscription may not outlive an activeConnection. For example, with lambdai@'s recent change, a RDS subscription will outlive a listener. when that happens, the update_on_demand_callback.cb_() end in undefined behavior.

Reply:
> No, I am saying a subscription may outlive an activeConnection. For example, with lambdai@'s recent change, a RDS subscription will outlive a listener. when that happens, the update_on_demand_callback.cb_() end in undefined behavior.